space_packet_parser/xtce/definitions.py

```diff
@@ -74,16 +74,18 @@ def __init__(
         author : Optional[str]
             Optional author name to include in XML when serializing.
         """
-        self._parameter_type_cache = {}
-        self._parameter_cache = {}
-        self._sequence_container_cache = {}
+        self.parameter_types = {}
+        self.parameters = {}
+        self.containers = {}
 
         # Populate the three caches for easy lookup later.
         if sequence_container_list:
             for sc in sequence_container_list:
-                self._parameter_type_cache.update({p.parameter_type.name: p.parameter_type for p in sc.entry_list})
-                self._parameter_cache.update({p.name: p for p in sc.entry_list})
-            self._sequence_container_cache.update({sc.name: sc for sc in sequence_container_list})
+                # TODO: entry_list can contain SequenceContainers too (parseables) so this needs an
+                #  isinstance check.
+                self.parameter_types.update({p.parameter_type.name: p.parameter_type for p in sc.entry_list})
+                self.parameters.update({p.name: p for p in sc.entry_list})
+            self.containers.update({sc.name: sc for sc in sequence_container_list})
```

Comment:
You are looping multiple times over the same lists, so maybe just do things explicitly with nested loops for simplicity?
```suggestion
                # entry_list can contain other containers, so filter those out and only add parameters
                for p in sc.entry_list if isinstance(p, parameters.Parameter):
                    if p.name not in self.parameters:
                        self.parameters[p.name] = p
                    if p.parameter_type.name not in self.parameter_types:
                        self.parameter_types[p.parameter_type.name] = p.parameter_type
                if sc.name not in self.containers:
                    self.containers[sc.name] = sc
```

Reply:
I actually already reworked this basically as you suggest, in order to raise exceptions for `Parameter`, `ParameterType`, and `SequenceContainer` objects of the same name that are equal (`a == b`) but are not the same object in memory (`a is b`).